### PR TITLE
de-mystify tests & run them in parallel

### PR DESCRIPTION
**Pull request by zramsay**
reduces the reliance on bash scripts (see https://github.com/tendermint/coding/issues/45) by putting various test pathways as `make` targets. This let's us use circle's parallelism feature to shorten test time and prepares for move to jenkins

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -62,13 +62,50 @@ draw_deps:
...
+	docker build -t tester -f ./test/docker/Dockerfile .
+
+### coverage, app, persistence, and libs tests
+test_cover:
+	# cleanup
+	bash ./test/test.sh
+	# run the go unit tests with coverage (in docker)
+	docker run --name run_test -t tester bash test/test_cover.sh
+	
+test_cover_ci:
```
Comment (zramsay) [inline Makefile]: the `*_ci` suffix is temporary and replaces some logic in the `test.sh` file. The won't be necessary when jenkins is setup

File: Makefile [hunk 2]
```diff
@@ -42,7 +42,7 @@ check_tools:
 get_tools:
 	@echo "--> Installing tools"
 	go get -u -v $(GOTOOLS)
-	@gometalinter.v2 --install
+	#@gometalinter.v2 --install
```
Comment (zramsay) [inline Makefile]: punt the linter until jenkins

File: circle.yml [hunk 3]
```diff
@@ -25,10 +25,12 @@ dependencies:
 
 test:
   override:
-    - cd "$PROJECT_PATH" && set -o pipefail && make test_integrations 2>&1 | tee test_integrations.log:
+   - case ${CIRCLE_NODE_INDEX} in 0) cd "$PROJECT_PATH" && make build_docker_test_image && make test_cover_ci;; 1) cd "$PROJECT_PATH" && make build_docker_test_image && make test_apps_ci;; 2) cd "$PROJECT_PATH" && make build_docker_test_image && make test_persistence_ci;; 3) cd "$PROJECT_PATH" && make build_docker_test_image && make test_p2p;; esac:
```
Comment (zramsay) [inline circle.yml]: ugly but works. `build_docker...` could be added as pre-run to the relevant make targets but that's gonna get re-worked with jenkins anyway

File: circle.yml [hunk 4]
```diff
@@ -25,10 +25,12 @@ dependencies:
...
-  post:
-    - cd "$PROJECT_PATH" && mv test_integrations.log "${CIRCLE_ARTIFACTS}"
-    - cd "$PROJECT_PATH" && bash <(curl -s https://codecov.io/bash) -f coverage.txt
-    - cd "$PROJECT_PATH" && mv coverage.txt "${CIRCLE_ARTIFACTS}"
-    - cd "$PROJECT_PATH" && cp test/logs/messages "${CIRCLE_ARTIFACTS}/docker_logs.txt"
+  #post:
+    #- cd "$PROJECT_PATH" && mv test_integrations.log "${CIRCLE_ARTIFACTS}"
+    #- cd "$PROJECT_PATH" && bash <(curl -s https://codecov.io/bash) -f coverage.txt
+    #- cd "$PROJECT_PATH" && mv coverage.txt "${CIRCLE_ARTIFACTS}"
+    #- cd "$PROJECT_PATH" && cp test/logs/messages "${CIRCLE_ARTIFACTS}/docker_logs.txt"
```
Comment (zramsay) [inline circle.yml]: not sure what/how to do with this with the parallelism

File: test/docker/Dockerfile [hunk 5]
```diff
@@ -25,6 +25,7 @@ ADD scripts scripts
 RUN bash scripts/install_abci_apps.sh
 
 # Now copy in the code
+# NOTE: this will overwrite whatever is in vendor/
```
Comment (zramsay) [inline test/docker/Dockerfile]: ...and kill any hope of determinism in the dockerfile

File: test/net/test.sh [hunk 6]
```diff
@@ -1,8 +0,0 @@
-#! /bin/bash
-set -eu
-
-# install mintnet, netmon, fetch network_testing
-bash test/net/setup.sh
-
-# start the testnet
-bash test/net/start.sh
```
Comment (zramsay) [inline test/net/test.sh]: all the mintnet stuff is deprecated. nothing useful in this dir

File: test/persist/test_failure_indices.sh [hunk 7]
```diff
@@ -24,7 +24,7 @@ function start_procs(){
     PID_DUMMY=$!
 
     # before starting tendermint, remove the rpc socket
-    rm $RPC_ADDR
+    rm -f $RPC_ADDR
```
Comment (zramsay) [inline test/persist/test_failure_indices.sh]: throws a warning on first run

File: test/run_test.sh [hunk 8]
```diff
@@ -1,26 +0,0 @@
...
-
-# run the persistence tests using bash
-bash test/persist/test.sh
-
-if [[ "$BRANCH" == "master" || $(echo "$BRANCH" | grep "release-") != "" ]]; then
-	echo ""
-	echo "* branch $BRANCH; testing libs"
-	# checkout every github.com/tendermint dir and run its tests
-	bash test/test_libs.sh
-fi
```
Comment (zramsay) [inline test/run_test.sh]: all moved into Makefile, to be consumed by Jenkins and users

File: test/test.sh [hunk 9]
```diff
@@ -23,38 +23,8 @@ echo "* [$(date +"%T")] starting rsyslog container"
 docker rm -f rsyslog || true
 docker run -d -v "$LOGS_DIR:/var/log/" -p 127.0.0.1:5514:514/udp --name rsyslog voxxit/rsyslog
 
-set +u
-if [[ "$SKIP_BUILD" == "" ]]; then
-	echo
-	echo "* [$(date +"%T")] building docker image"
-	bash "$DIR/docker/build.sh"
-fi
+pwd
```
Comment (zramsay) [inline test/test.sh]: copied over from somewhere ... ?

File: test/test.sh [hunk 10]
```diff
@@ -23,38 +23,8 @@ echo "* [$(date +"%T")] starting rsyslog container"
...
-if [[ $(echo "$BRANCH" | grep "release-") != "" ]]; then
-	echo
-	echo "TODO: run network tests"
-	#echo "* branch $BRANCH; running mintnet/netmon throughput benchmark"
-	# TODO: replace mintnet
-	#bash "$DIR/net/test.sh"
-fi
+#TODO
+## copy the coverage results out of docker container
+#docker cp run_test:/go/src/github.com/tendermint/tendermint/coverage.txt .
```
Comment (zramsay) [inline test/test.sh]: also to be dealt with in circle.yml  - or just do it right in jenkins once

File: test/test.sh [hunk 11]
```diff
@@ -23,38 +23,8 @@ echo "* [$(date +"%T")] starting rsyslog container"
...
+pwd
 
-echo
-echo "* [$(date +"%T")] running go tests and app tests in docker container"
-# sometimes its helpful to mount the local test folder
-# -v $DIR:/go/src/github.com/tendermint/tendermint/test
-if [[ "$CIRCLECI" == true ]]; then
-	docker run --name run_test -e CIRCLECI=true -t tester bash test/run_test.sh
-else
-	docker run --name run_test -t tester bash test/run_test.sh
```
Comment (zramsay) [inline test/test.sh]: moved and split to Makefile

File: test/test.sh [hunk 12]
```diff
@@ -23,38 +23,8 @@ echo "* [$(date +"%T")] starting rsyslog container"
 docker rm -f rsyslog || true
 docker run -d -v "$LOGS_DIR:/var/log/" -p 127.0.0.1:5514:514/udp --name rsyslog voxxit/rsyslog
 
-set +u
-if [[ "$SKIP_BUILD" == "" ]]; then
-	echo
-	echo "* [$(date +"%T")] building docker image"
-	bash "$DIR/docker/build.sh"
```
Comment (zramsay) [inline test/test.sh]: in makefile

File: circle.yml (hunk 4 above)
Comment (ebuchman) [reply circle.yml]: this is ditching the coverage and the logs. We need to keep both ... I don't know how worth it it is to get this stuff done on Circle if we're only going to move to Jenkins ...

File: Makefile [hunk 13]
```diff
@@ -85,6 +122,15 @@ vagrant_test:
 	vagrant ssh -c 'make test_race'
 	vagrant ssh -c 'make test_integrations'
 
+### go tests without docker
+test_unit:
```
Comment (melekes) [inline Makefile]: why not just `test`?

File: Makefile [hunk 14]
```diff
@@ -62,13 +62,50 @@ draw_deps:
...
-	@echo "--> Running go test"
-	@go test $(PACKAGES)
+## required to be run first by most tests
+build_docker_test_image:
+	docker build -t tester -f ./test/docker/Dockerfile .
+
+### coverage, app, persistence, and libs tests
+test_cover:
+	# cleanup
+	bash ./test/test.sh
```
Comment (melekes) [inline Makefile]: lets rename to `./test/cleanup.sh`

File: test/test.sh (hunk 9 above)
Comment (melekes) [reply test/test.sh]: lets remove this too

File: Makefile (hunk 13 above)
Comment (zramsay) [reply Makefile]: `test` would imply all the tests

**Comment by zramsay** [comment]
merging to jenkins branch

**Pull request merged**